hibernate-core/src/main/java/org/hibernate/internal/util/ReflectHelper.java

```diff
@@ -49,6 +51,9 @@
 	private static final Method OBJECT_EQUALS;
 	private static final Method OBJECT_HASHCODE;
 
+	private static final Map<Class<?>, Method[]> METHODS_BY_CLASS = new HashMap<>();
+	private static final Map<Class<?>, Field[]> FIELDS_BY_CLASS = new HashMap<>();
```

Comment:
So this is not possible for us to, unfortunately. We can't have the Hibernate classloader hold a "hard reference" to a class instance of the application: many application servers woul start leaking memory, as the Hibernate module is long-lived while applications (and user classes) are often undeployed, redeployed, etc..  sometimes many times.

What we can do is to introuce a new class type `IntrospectedClass` (or anything) to store together: the user class, its methods, its fields. The methods array and the fields array could be lazily initialized.
Then you refactor this class to consistently use that new `IntropectedClass` instead of the current Class.

If I'm unclear let me know, I'm happy to show you what I mean :)

Reply:
Hi @Sanne,

sorry for the late reply, had to focus my creativity on real (as in paid) work for the last few weeks.

I gave your suggestion a shot in https://github.com/Postremus/hibernate-orm/commit/7d02e58948860fbe77d3a9c8af038104cf04a71d.

I quickly noticed though that, as far as I understand, this won't work. `IntrospectedClass` is a clean idea to solve the problem of hard references on class instances. 
However, the current structure of ORM does not lend itself to this concept.

Using `IntrospectedClass` would also mean using it in all call sites and their call sites in turn. This would result in way to many changes, and also needs changes in spi interfaces (e.g. PropertyAccess, Getter, Setter).
Otherwise `IntrospectedClass` is always for each field access, rendering it useless.

I have switched the commit order around abit (I love interactive rebasing) - this commit is now the last. I am okay with simply dropping it.
ORM 6 is dropping in the next few weeks anyway, where work could be started on using jandex instead of reflection.